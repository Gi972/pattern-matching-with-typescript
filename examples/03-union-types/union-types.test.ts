import { Argument, matchArgument } from './union-types';

describe('Union Types: ArgumentPattern', () => {
  it('should match a string with String(string)', () => {
    const argument: Argument = 'Hello World.';
    const stringSpy = jest.fn();
    const booleanSpy = jest.fn();
    const numberSpy = jest.fn();

    matchArgument({
      String: stringSpy,
      Boolean: booleanSpy,
      Number: numberSpy
    })(argument);

    expect(stringSpy).toHaveBeenCalledWith(argument);
    expect(booleanSpy).not.toHaveBeenCalled();
    expect(numberSpy).not.toHaveBeenCalled();
  });

  it('should match a boolean with Boolean(boolean)', () => {
    const argument: Argument = true;
    const stringSpy = jest.fn();
    const booleanSpy = jest.fn();
    const numberSpy = jest.fn();

    matchArgument({
      String: stringSpy,
      Boolean: booleanSpy,
      Number: numberSpy
    })(argument);

    expect(stringSpy).not.toHaveBeenCalled();
    expect(booleanSpy).toHaveBeenCalledWith(argument);
    expect(numberSpy).not.toHaveBeenCalled();
  });

  it('should match null with Null()', () => {
    const argument: Argument = 42;
    const stringSpy = jest.fn();
    const booleanSpy = jest.fn();
    const numberSpy = jest.fn();

    matchArgument({
      String: stringSpy,
      Boolean: booleanSpy,
      Number: numberSpy
    })(argument);

    expect(stringSpy).not.toHaveBeenCalled();
    expect(booleanSpy).not.toHaveBeenCalled();
    expect(numberSpy).toHaveBeenCalledWith(argument);
  });

  it('should throw an error when trying to match undefined', () => {
    const argument: Argument = undefined;
    const stringSpy = jest.fn();
    const booleanSpy = jest.fn();
    const numberSpy = jest.fn();

    expect(() => {
      matchArgument({
        String: stringSpy,
        Boolean: booleanSpy,
        Number: numberSpy
      })(argument);
    }).toThrowErrorMatchingSnapshot();
  });
});
