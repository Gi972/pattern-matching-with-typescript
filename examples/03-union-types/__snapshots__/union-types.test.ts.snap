// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Union Types: ArgumentPattern should throw an error when trying to match undefined 1`] = `"matchArgument: Could not match type undefined"`;
